File: client/src/app/services/consultationService/consultation-service.ts
import { Injectable } from '@angular/core';
import { ApiService } from '../../API/api-service';
import { ConsultationApi } from '../../API/consultationAPI/consultation-api';
import { Router } from '@angular/router';
import { BehaviorSubject, Observable } from 'rxjs';
import { tap, catchError } from 'rxjs/operators';

@Injectable({
  providedIn: 'root',
})
export class ConsultationService {
  constructor(
    private api: ApiService,
    private consultationApi: ConsultationApi,
    private router: Router
  ) {}

  private listConsultationDataSubject = new BehaviorSubject<any | null>(null);
  listConsultationData$ = this.listConsultationDataSubject.asObservable();

  // ==================== Legacy Methods (Backward Compatibility) ====================

  getConsultationRequests() {
    if (!this.listConsultationDataSubject.getValue()) {
      this.getListConsultationRequestsData();
    }
    console.log('getConsultationRequests');
    console.log(this.listConsultationDataSubject.getValue());
    return this.listConsultationData$;
  }

  getListConsultationRequestsData() {
    console.log('🔵 Fetching consultation requests from API...');
    this.api.getConsultationRequestsByConsultant().subscribe({
      next: (res: any) => {
        console.log('✅ CONSULTANT::getListConsultationRequestsData::SUCCESS: ', res);
        this.listConsultationDataSubject.next(res.data);
      },
      error: (err: any) => {
        console.error('❌ CONSULTANT::getListConsultationRequestsData::ERROR: ', err);
      },
    });
  }

  acceptConsultationRequest(consultationId: any) {
    console.log('🟡 Accepting consultation request:', consultationId);
    this.api.acceptConsultationRequest(consultationId).subscribe({
      next: (res: any) => {
        console.log('✅ CONSULTANT::acceptConsultationRequest::SUCCESS: ', res);

        // Refresh data immediately after success
        setTimeout(() => {
          this.getListConsultationRequestsData();
        }, 500); // Small delay to ensure backend updates
      },
      error: (err: any) => {
        console.error('❌ CONSULTANT::acceptConsultationRequest::ERROR: ', err);
      },
    });
  }

  rejectConsultationRequest(consultationId: any) {
    console.log('🟡 Rejecting consultation request:', consultationId);
    this.api.rejectConsultationRequest(consultationId).subscribe({
      next: (res: any) => {
        console.log('✅ CONSULTANT::rejectConsultationRequest::SUCCESS: ', res);

        // Refresh data immediately after success
        setTimeout(() => {
          this.getListConsultationRequestsData();
        }, 500); // Small delay to ensure backend updates
      },
      error: (err: any) => {
        console.error('❌ CONSULTANT::rejectConsultationRequest::ERROR: ', err);
      },
    });
  }

  // SCHEDULE FARM VISIT
  scheduleFarmVisit(farmVisitData: any, consultationId: number) {
    console.log('🟡 Scheduling farm visit with data:', farmVisitData);
    this.api.scheduleFarmVisit(farmVisitData, consultationId).subscribe({
      next: (res: any) => {
        console.log('✅ CONSULTANT::scheduleFarmVisit::SUCCESS: ', res);

        setTimeout(() => {
          this.getListConsultationRequestsData();
        }, 500); // Small delay to ensure backend updates
      },
      error: (err: any) => {
        console.error('❌ CONSULTANT::scheduleFarmVisit::ERROR: ', err);
      },
    });
  }

  // CREATE CONSULTATION REPORT
  createConsultationReport(reportData: any) {
    console.log('🟡 Creating consultation report with data:', reportData);
    this.api.createConsultationReport(reportData).subscribe({
      next: (res: any) => {
        console.log('✅ CONSULTANT::createConsultationReport::SUCCESS: ', res);

        setTimeout(() => {
          this.getListConsultationRequestsData();
        }, 500); // Small delay to ensure backend updates
      },
      error: (err: any) => {
        console.error('❌ CONSULTANT::createConsultationReport::ERROR: ', err);
      },
    });
  }

  // ==================== New Consultation API Methods ====================

  /**
   * Get consultation by ID
   */
  getConsultationById(id: number): Observable<any> {
    console.log(`🔵 Fetching consultation with ID: ${id}`);
    return this.consultationApi.getConsultationById(id).pipe(
      tap((res: any) => console.log('✅ getConsultationById::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getConsultationById::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Get all consultations for a farmer
   */
  getConsultationsByFarmerId(farmerId: number): Observable<any> {
    console.log(`🔵 Fetching consultations for farmer ID: ${farmerId}`);
    return this.consultationApi.getConsultationsByFarmerId(farmerId).pipe(
      tap((res: any) => console.log('✅ getConsultationsByFarmerId::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getConsultationsByFarmerId::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Get all consultations for a consultant
   */
  getConsultationsByConsultantId(consultantId: number): Observable<any> {
    console.log(`🔵 Fetching consultations for consultant ID: ${consultantId}`);
    return this.consultationApi.getConsultationsByConsultantId(consultantId).pipe(
      tap((res: any) => console.log('✅ getConsultationsByConsultantId::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getConsultationsByConsultantId::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Get consultations by crop
   */
  getConsultationsByCropId(cropId: number): Observable<any> {
    console.log(`🔵 Fetching consultations for crop ID: ${cropId}`);
    return this.consultationApi.getConsultationsByCropId(cropId).pipe(
      tap((res: any) => console.log('✅ getConsultationsByCropId::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getConsultationsByCropId::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Get recent consultations
   */
  getRecentConsultations(): Observable<any> {
    console.log('🔵 Fetching recent consultations');
    return this.consultationApi.getRecentConsultations().pipe(
      tap((res: any) => console.log('✅ getRecentConsultations::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getRecentConsultations::ERROR:', err);
        throw err;
      })
    );
  }

  // ==================== Status-Based Methods ====================

  /**
   * Get consultations by status
   */
  getConsultationsByStatus(status: string): Observable<any> {
    console.log(`🔵 Fetching consultations with status: ${status}`);
    return this.consultationApi.getConsultationsByStatus(status).pipe(
      tap((res: any) => console.log('✅ getConsultationsByStatus::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getConsultationsByStatus::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Get farmer's consultations by status
   */
  getConsultationsByFarmerAndStatus(farmerId: number, status: string): Observable<any> {
    console.log(`🔵 Fetching consultations for farmer ${farmerId} with status: ${status}`);
    return this.consultationApi.getConsultationsByFarmerAndStatus(farmerId, status).pipe(
      tap((res: any) => console.log('✅ getConsultationsByFarmerAndStatus::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getConsultationsByFarmerAndStatus::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Get consultant's consultations by status
   */
  getConsultationsByConsultantAndStatus(consultantId: number, status: string): Observable<any> {
    console.log(`🔵 Fetching consultations for consultant ${consultantId} with status: ${status}`);
    return this.consultationApi.getConsultationsByConsultantAndStatus(consultantId, status).pipe(
      tap((res: any) => console.log('✅ getConsultationsByConsultantAndStatus::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getConsultationsByConsultantAndStatus::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Get consultations with multiple statuses
   */
  getConsultationsByMultipleStatuses(statuses: string[]): Observable<any> {
    console.log(`🔵 Fetching consultations with statuses: ${statuses.join(', ')}`);
    return this.consultationApi.getConsultationsByMultipleStatuses(statuses).pipe(
      tap((res: any) => console.log('✅ getConsultationsByMultipleStatuses::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getConsultationsByMultipleStatuses::ERROR:', err);
        throw err;
      })
    );
  }

  // ==================== Status Management Methods ====================

  /**
   * Approve a consultation
   */
  approveConsultation(id: number): Observable<any> {
    console.log(`🟡 Approving consultation ID: ${id}`);
    return this.consultationApi.approveConsultation(id).pipe(
      tap((res: any) => {
        console.log('✅ approveConsultation::SUCCESS:', res);
        this.refreshConsultationData();
      }),
      catchError((err: any) => {
        console.error('❌ approveConsultation::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Reject a consultation
   */
  rejectConsultation(id: number): Observable<any> {
    console.log(`🟡 Rejecting consultation ID: ${id}`);
    return this.consultationApi.rejectConsultation(id).pipe(
      tap((res: any) => {
        console.log('✅ rejectConsultation::SUCCESS:', res);
        this.refreshConsultationData();
      }),
      catchError((err: any) => {
        console.error('❌ rejectConsultation::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Complete a consultation
   */
  completeConsultation(id: number): Observable<any> {
    console.log(`🟡 Completing consultation ID: ${id}`);
    return this.consultationApi.completeConsultation(id).pipe(
      tap((res: any) => {
        console.log('✅ completeConsultation::SUCCESS:', res);
        this.refreshConsultationData();
      }),
      catchError((err: any) => {
        console.error('❌ completeConsultation::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Update consultation status
   */
  updateConsultationStatus(id: number, status: string): Observable<any> {
    console.log(`🟡 Updating consultation ID ${id} status to: ${status}`);
    return this.consultationApi.updateConsultationStatus(id, status).pipe(
      tap((res: any) => {
        console.log('✅ updateConsultationStatus::SUCCESS:', res);
        this.refreshConsultationData();
      }),
      catchError((err: any) => {
        console.error('❌ updateConsultationStatus::ERROR:', err);
        throw err;
      })
    );
  }

  // ==================== Update Methods ====================

  /**
   * Update consultation topic and description
   */
  updateConsultation(id: number, topic: string, description: string): Observable<any> {
    console.log(`🟡 Updating consultation ID: ${id}`);
    return this.consultationApi.updateConsultation(id, topic, description).pipe(
      tap((res: any) => {
        console.log('✅ updateConsultation::SUCCESS:', res);
        this.refreshConsultationData();
      }),
      catchError((err: any) => {
        console.error('❌ updateConsultation::ERROR:', err);
        throw err;
      })
    );
  }

  // ==================== Statistics Methods ====================

  /**
   * Get total consultation count for farmer
   */
  getFarmerConsultationCount(farmerId: number): Observable<any> {
    console.log(`🔵 Fetching consultation count for farmer ID: ${farmerId}`);
    return this.consultationApi.getFarmerConsultationCount(farmerId).pipe(
      tap((res: any) => console.log('✅ getFarmerConsultationCount::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getFarmerConsultationCount::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Get total consultation count for consultant
   */
  getConsultantConsultationCount(consultantId: number): Observable<any> {
    console.log(`🔵 Fetching consultation count for consultant ID: ${consultantId}`);
    return this.consultationApi.getConsultantConsultationCount(consultantId).pipe(
      tap((res: any) => console.log('✅ getConsultantConsultationCount::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getConsultantConsultationCount::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Get consultation count by status
   */
  getConsultationCountByStatus(status: string): Observable<any> {
    console.log(`🔵 Fetching consultation count for status: ${status}`);
    return this.consultationApi.getConsultationCountByStatus(status).pipe(
      tap((res: any) => console.log('✅ getConsultationCountByStatus::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ getConsultationCountByStatus::ERROR:', err);
        throw err;
      })
    );
  }

  // ==================== Validation Methods ====================

  /**
   * Check if active consultation exists between farmer and consultant
   */
  hasActiveConsultation(farmerId: number, consultantId: number): Observable<any> {
    console.log(
      `🔵 Checking active consultation between farmer ${farmerId} and consultant ${consultantId}`
    );
    return this.consultationApi.hasActiveConsultation(farmerId, consultantId).pipe(
      tap((res: any) => console.log('✅ hasActiveConsultation::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ hasActiveConsultation::ERROR:', err);
        throw err;
      })
    );
  }

  /**
   * Validate if consultation can be created
   */
  canCreateConsultation(farmerId: number, consultantId: number): Observable<any> {
    console.log(
      `🔵 Validating if consultation can be created between farmer ${farmerId} and consultant ${consultantId}`
    );
    return this.consultationApi.canCreateConsultation(farmerId, consultantId).pipe(
      tap((res: any) => console.log('✅ canCreateConsultation::SUCCESS:', res)),
      catchError((err: any) => {
        console.error('❌ canCreateConsultation::ERROR:', err);
        throw err;
      })
    );
  }

  // ==================== Delete Method ====================

  /**
   * Delete a consultation
   */
  deleteConsultation(id: number): Observable<any> {
    console.log(`🟡 Deleting consultation ID: ${id}`);
    return this.consultationApi.deleteConsultation(id).pipe(
      tap((res: any) => {
        console.log('✅ deleteConsultation::SUCCESS:', res);
        this.refreshConsultationData();
      }),
      catchError((err: any) => {
        console.error('❌ deleteConsultation::ERROR:', err);
        throw err;
      })
    );
  }

  // ==================== Helper Methods ====================

  /**
   * Refresh consultation data after updates
   */
  private refreshConsultationData(): void {
    setTimeout(() => {
      this.getListConsultationRequestsData();
    }, 500);
  }
}

File: server/src/main/java/com/server/service/ConsultationService.java
package com.server.service;

import com.server.dto.ConsultationRequestDTO;
import com.server.entity.*;
import com.server.enumeration.ConsultationRequestStatus;
import com.server.repository.ConsultationRepository;
import jakarta.transaction.Transactional;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
@Slf4j
public class ConsultationService {
    @Autowired
    private ConsultationRepository consultationRepository;
    @Autowired
    private AddressService addressService;

    @Autowired
    private ConsultantService consultantService;
    @Autowired
    private CropService cropService;
    @Autowired
    private SimpleChatService chatService;

    @Transactional
    public Optional<Consultation> createConsultationRequest(
            Farmer farmer,
            ConsultationRequestDTO request) {
        log.info("Creating consultation request for farmer: {}", farmer.getEmail());

        // 1. Find consultant using email (NOT ID)
        Consultant consultant = consultantService.getConsultantByUsername(request.getConsultantEmail())
                .orElseThrow(() -> new RuntimeException(
                        "Consultant not found with username: " + request.getConsultantEmail()));

        log.info("Found consultant: {}", consultant.getEmail());

        // 2. Save or get crop
        Crop crop = new Crop();
        crop.setName(request.getCrop().getName());
        crop.setCategory(request.getCrop().getCategory());
        crop.setType(request.getCrop().getType());
        crop.setDescription(request.getCrop().getDescription());
        crop.setCreatedAt(LocalDateTime.now());
        crop = cropService.save(crop)
                .orElseThrow(() -> new RuntimeException("Failed to save or fetch crop"));
        /*
        
         */
        log.info("Crop linked to consultation: {}", crop.getName());

        // 3. Create Address (Farm Address)
        Address address;
        if (!request.getUseExistingAddress()) {
            log.info("Creating new farm address for consultation");
            address = new Address();
            address.setStreet(request.getFarmAddress().getStreet());
            address.setCity(request.getFarmAddress().getCity());
            address.setState(request.getFarmAddress().getState());
            address.setPinCode(request.getFarmAddress().getPinCode());
            address.setCountry(request.getFarmAddress().getCountry());
            address.setLatitude(request.getFarmAddress().getLatitude());
            address.setLongitude(request.getFarmAddress().getLongitude());
        } else {
            log.info("Using existing farm address for consultation");
            address = farmer.getAddress();
        }
        // Persist address manually (otherwise it will not save)
        Address savedAddress = addressService.save(address)
                .orElseThrow(() -> new RuntimeException("Failed to save or fetch Address"));
        log.info("Saved farm address: {}", savedAddress.getId());

        // 4. Build Consultation Entity
        Consultation consultation = new Consultation();
        consultation.setFarmer(farmer);
        consultation.setConsultant(consultant);
        consultation.setCrop(crop);
        consultation.setFarmAddress(savedAddress);
        consultation.setTopic(request.getTopic());
        consultation.setDescription(request.getDescription());
        consultation.setConsultationRequestStatus(ConsultationRequestStatus.PENDING);
        consultation.setCreatedAt(LocalDateTime.now());
        consultation.setUpdatedAt(LocalDateTime.now());

        Consultation savedConsultation = consultationRepository.save(consultation);
        log.info("Consultation request created with id: {}", savedConsultation.getId());

        return Optional.of(savedConsultation);
    }

    // ==================== Retrieval Methods ====================

    public Optional<Consultation> getConsultationById(Long id) {
        log.info("Fetching consultation request with id: {}", id);
        return consultationRepository.findById(id);
    }

    public List<Consultation> getConsultationsByFarmer(Farmer farmer) {
        log.info("Fetching consultations for farmer: {}", farmer.getEmail());
        return consultationRepository.findByFarmerId(farmer.getId());
    }

    public List<Consultation> getConsultationsByFarmerId(Long farmerId) {
        log.info("Fetching consultations for farmer id: {}", farmerId);
        return consultationRepository.findByFarmerId(farmerId);
    }

    public List<Consultation> getConsultationsByConsultantId(Long consultantId) {
        log.info("Fetching consultations for consultant id: {}", consultantId);
        return consultationRepository.findByConsultantId(consultantId);
    }

    public List<Consultation> getConsultationsByCropId(Long cropId) {
        log.info("Fetching consultations for crop id: {}", cropId);
        return consultationRepository.findByCropId(cropId);
    }

    // ==================== Status-based Retrieval ====================

    public List<Consultation> getConsultationsByStatus(ConsultationRequestStatus status) {
        log.info("Fetching consultations with status: {}", status);
        return consultationRepository.findByConsultationRequestStatus(status);
    }

    public List<Consultation> getConsultationsByFarmerAndStatus(Long farmerId, ConsultationRequestStatus status) {
        log.info("Fetching consultations for farmer {} with status: {}", farmerId, status);
        return consultationRepository.findByFarmerIdAndConsultationRequestStatus(farmerId, status);
    }

    public List<Consultation> getConsultationsByConsultantAndStatus(Long consultantId,
            ConsultationRequestStatus status) {
        log.info("Fetching consultations for consultant {} with status: {}", consultantId, status);
        return consultationRepository.findByConsultantIdAndConsultationRequestStatus(consultantId, status);
    }

    public List<Consultation> getConsultationsByStatusIn(List<ConsultationRequestStatus> statuses) {
        log.info("Fetching consultations with statuses: {}", statuses);
        return consultationRepository.findByStatusIn(statuses);
    }

    public List<Consultation> getConsultationsByFarmerAndStatusIn(Long farmerId,
            List<ConsultationRequestStatus> statuses) {
        log.info("Fetching consultations for farmer {} with statuses: {}", farmerId, statuses);
        return consultationRepository.findByFarmerIdAndStatusIn(farmerId, statuses);
    }

    public List<Consultation> getConsultationsByConsultantAndStatusIn(Long consultantId,
            List<ConsultationRequestStatus> statuses) {
        log.info("Fetching consultations for consultant {} with statuses: {}", consultantId, statuses);
        return consultationRepository.findByConsultantIdAndStatusIn(consultantId, statuses);
    }

    // ==================== Date-based Retrieval ====================

    public List<Consultation> getRecentConsultations() {
        log.info("Fetching top 10 recent consultations");
        return consultationRepository.findTop10ByOrderByCreatedAtDesc();
    }

    public List<Consultation> getConsultationsByDateRange(LocalDateTime startDate, LocalDateTime endDate) {
        log.info("Fetching consultations between {} and {}", startDate, endDate);
        return consultationRepository.findByCreatedAtBetween(startDate, endDate);
    }

    public List<Consultation> getConsultationsAfterDate(LocalDateTime date) {
        log.info("Fetching consultations created after {}", date);
        return consultationRepository.findByCreatedAtAfter(date);
    }

    public List<Consultation> getConsultationsByStatusAndDateAfter(ConsultationRequestStatus status,
            LocalDateTime date) {
        log.info("Fetching consultations with status {} created after {}", status, date);
        return consultationRepository.findByConsultationRequestStatusAndCreatedAtAfter(status, date);
    }

    // ==================== Status Management Methods ====================

    @Transactional
    public Optional<Consultation> approveConsultation(Long consultationId) {
        log.info("Approving consultation with id: {}", consultationId);

        Optional<Consultation> consultationOpt = consultationRepository.findById(consultationId);
        if (consultationOpt.isEmpty()) {
            log.error("Consultation not found with id: {}", consultationId);
            return Optional.empty();
        }

        Consultation consultation = consultationOpt.get();

        if (consultation.getConsultationRequestStatus() != ConsultationRequestStatus.PENDING) {
            log.warn("Cannot approve consultation {} - current status is {}",
                    consultationId, consultation.getConsultationRequestStatus());
            throw new RuntimeException("Only PENDING consultations can be approved");
        }

        consultation.setConsultationRequestStatus(ConsultationRequestStatus.APPROVED);
        consultation.setUpdatedAt(LocalDateTime.now());

        Consultation updated = consultationRepository.save(consultation);
        log.info("Consultation {} approved successfully", consultationId);

        // Auto-create chat room using SimpleChatService
        try {
            chatService.createChatRoom(updated);
            log.info("Chat room created for consultation: {}", consultationId);
        } catch (Exception e) {
            log.error("Failed to create chat room for consultation {}: {}", consultationId, e.getMessage());
            // Don't fail the transaction if chat creation fails, just log it
        }

        return Optional.of(updated);
    }

    @Transactional
    public Optional<Consultation> rejectConsultation(Long consultationId) {
        log.info("Rejecting consultation with id: {}", consultationId);

        Optional<Consultation> consultationOpt = consultationRepository.findById(consultationId);
        if (consultationOpt.isEmpty()) {
            log.error("Consultation not found with id: {}", consultationId);
            return Optional.empty();
        }

        Consultation consultation = consultationOpt.get();

        if (consultation.getConsultationRequestStatus() != ConsultationRequestStatus.PENDING) {
            log.warn("Cannot reject consultation {} - current status is {}",
                    consultationId, consultation.getConsultationRequestStatus());
            throw new RuntimeException("Only PENDING consultations can be rejected");
        }

        consultation.setConsultationRequestStatus(ConsultationRequestStatus.REJECTED);
        consultation.setUpdatedAt(LocalDateTime.now());

        Consultation updated = consultationRepository.save(consultation);
        log.info("Consultation {} rejected successfully", consultationId);

        return Optional.of(updated);
    }

    @Transactional
    public Optional<Consultation> completeConsultation(Long consultationId) {
        log.info("Completing consultation with id: {}", consultationId);

        Optional<Consultation> consultationOpt = consultationRepository.findById(consultationId);
        if (consultationOpt.isEmpty()) {
            log.error("Consultation not found with id: {}", consultationId);
            return Optional.empty();
        }

        Consultation consultation = consultationOpt.get();

        if (consultation.getConsultationRequestStatus() != ConsultationRequestStatus.APPROVED) {
            log.warn("Cannot complete consultation {} - current status is {}",
                    consultationId, consultation.getConsultationRequestStatus());
            throw new RuntimeException("Only APPROVED consultations can be completed");
        }

        consultation.setConsultationRequestStatus(ConsultationRequestStatus.COMPLETED);
        consultation.setUpdatedAt(LocalDateTime.now());
        consultation.setClosedAt(LocalDateTime.now());

        Consultation updated = consultationRepository.save(consultation);
        log.info("Consultation {} completed successfully", consultationId);

        return Optional.of(updated);
    }

    @Transactional
    public Optional<Consultation> updateConsultationStatus(Long consultationId, ConsultationRequestStatus newStatus) {
        log.info("Updating consultation {} status to {}", consultationId, newStatus);

        Optional<Consultation> consultationOpt = consultationRepository.findById(consultationId);
        if (consultationOpt.isEmpty()) {
            log.error("Consultation not found with id: {}", consultationId);
            return Optional.empty();
        }

        Consultation consultation = consultationOpt.get();
        consultation.setConsultationRequestStatus(newStatus);
        consultation.setUpdatedAt(LocalDateTime.now());

        if (newStatus == ConsultationRequestStatus.COMPLETED) {
            consultation.setClosedAt(LocalDateTime.now());
        }

        Consultation updated = consultationRepository.save(consultation);
        log.info("Consultation {} status updated to {}", consultationId, newStatus);

        return Optional.of(updated);
    }

    // ==================== Statistics Methods ====================

    public Long getConsultationCountByFarmer(Long farmerId) {
        log.info("Counting consultations for farmer: {}", farmerId);
        return consultationRepository.countByFarmerId(farmerId);
    }

    public Long getConsultationCountByConsultant(Long consultantId) {
        log.info("Counting consultations for consultant: {}", consultantId);
        return consultationRepository.countByConsultantId(consultantId);
    }

    public Long getConsultationCountByStatus(ConsultationRequestStatus status) {
        log.info("Counting consultations with status: {}", status);
        return consultationRepository.countByConsultationRequestStatus(status);
    }

    public Long getConsultationCountByConsultantAndStatus(Long consultantId, ConsultationRequestStatus status) {
        log.info("Counting consultations for consultant {} with status: {}", consultantId, status);
        return consultationRepository.countByConsultantIdAndConsultationRequestStatus(consultantId, status);
    }

    public Long getConsultationCountByFarmerAndStatus(Long farmerId, ConsultationRequestStatus status) {
        log.info("Counting consultations for farmer {} with status: {}", farmerId, status);
        Long count = (long) consultationRepository.findByFarmerIdAndConsultationRequestStatus(farmerId, status).size();
        return count;
    }

    public Long getConsultationCountAfterDate(LocalDateTime date) {
        log.info("Counting consultations created after: {}", date);
        return consultationRepository.countByCreatedAtAfter(date);
    }

    public Long getConsultationCountByCrop(Long cropId) {
        log.info("Counting consultations for crop: {}", cropId);
        return consultationRepository.countByCropId(cropId);
    }

    // ==================== Validation Methods ====================

    public boolean hasActiveConsultation(Long farmerId, Long consultantId) {
        log.info("Checking for active consultation between farmer {} and consultant {}", farmerId, consultantId);
        return consultationRepository.existsByFarmerIdAndConsultantIdAndConsultationRequestStatus(
                farmerId, consultantId, ConsultationRequestStatus.APPROVED);
    }

    public boolean canCreateConsultation(Long farmerId, Long consultantId) {
        log.info("Validating if farmer {} can create consultation with consultant {}", farmerId, consultantId);
        // Check if there's already a pending consultation
        boolean hasPending = consultationRepository.existsByFarmerIdAndConsultantIdAndConsultationRequestStatus(
                farmerId, consultantId, ConsultationRequestStatus.PENDING);

        if (hasPending) {
            log.warn("Farmer {} already has a pending consultation with consultant {}", farmerId, consultantId);
            return false;
        }

        return true;
    }

    // ==================== Farm Visit Management ====================

    @Transactional
    public Optional<Consultation> addFarmVisitToConsultation(Consultation consultation, Farmvisit farmvisit) {
        log.info("Adding farm visit to consultation with id: {}", consultation.getId());
        consultation.getFarmVisits().add(farmvisit);
        Consultation updatedConsultation = consultationRepository.save(consultation);
        log.info("Farm visit added to consultation with id: {}", consultation.getId());
        return Optional.of(updatedConsultation);
    }

    @Transactional
    public Optional<Consultation> addFarmVisitById(Long consultationId, Farmvisit farmvisit) {
        log.info("Adding farm visit to consultation id: {}", consultationId);

        Optional<Consultation> consultationOpt = consultationRepository.findById(consultationId);
        if (consultationOpt.isEmpty()) {
            log.error("Consultation not found with id: {}", consultationId);
            return Optional.empty();
        }

        Consultation consultation = consultationOpt.get();
        consultation.getFarmVisits().add(farmvisit);
        Consultation updated = consultationRepository.save(consultation);
        log.info("Farm visit added to consultation {}", consultationId);

        return Optional.of(updated);
    }

    // ==================== Update Methods ====================

    @Transactional
    public Optional<Consultation> updateConsultation(Long consultationId, String topic, String description) {
        log.info("Updating consultation {}", consultationId);

        Optional<Consultation> consultationOpt = consultationRepository.findById(consultationId);
        if (consultationOpt.isEmpty()) {
            log.error("Consultation not found with id: {}", consultationId);
            return Optional.empty();
        }

        Consultation consultation = consultationOpt.get();

        if (topic != null && !topic.isEmpty()) {
            consultation.setTopic(topic);
        }
        if (description != null && !description.isEmpty()) {
            consultation.setDescription(description);
        }
        consultation.setUpdatedAt(LocalDateTime.now());

        Consultation updated = consultationRepository.save(consultation);
        log.info("Consultation {} updated successfully", consultationId);

        return Optional.of(updated);
    }

    // ==================== Delete Methods ====================

    @Transactional
    public boolean deleteConsultation(Long consultationId) {
        log.info("Deleting consultation {}", consultationId);

        if (!consultationRepository.existsById(consultationId)) {
            log.error("Consultation not found with id: {}", consultationId);
            return false;
        }

        consultationRepository.deleteById(consultationId);
        log.info("Consultation {} deleted successfully", consultationId);
        return true;
    }
}
